### ISSUE TITLE
[QC] How to cover private methods

### ISSUE THREAD

**Issue by dayo09** [Samsung/ONE-vscode#1095]
## What?

Even if private methods are covered internally by the module's public api, they are counted as uncovered lines.

For example,
current mocha coverage report says ConfigObjects line 150-184 and 193-228 are uncovered. However, they're the private functions that ConfigObject create function calls.

File                                         | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                             
---------------------------------------------|---------|----------|---------|---------|-----------------------------------------------                                        
  ConfigObject.ts                            |      30 |     12.5 |    9.09 |   31.57 | 60-94,115,135,150-184,193-228  

```typescript
  private static parseBaseModels = (filePath: string, iniObj: object): Artifact[] => {
/// <------------- (Line 150) From here, the code is considered not to have been covered. 

    const dir = path.dirname(filePath);

    let locatorRunner = new LocatorRunner();

    locatorRunner.register({
      artifactAttr: {ext: '.tflite', icon: new vscode.ThemeIcon('symbol-variable')},
      locator: new Locator((value: string) => LocatorRunner.searchWithExt('.tflite', value))
    });
...
  };
```

## Why?

Javascript doesn't provide method to test private functions. It only expose functions which are exported by modules.

**CONCERN** Do we really need to test private functions? 
IDK. IMO, they are better if tested, but it's just my opinion. Anyways I just searched how to unit-test them, for our test coverage, for the upcoming release.

### How to cover private functions?

Reference: [:link:](https://stackoverflow.com/questions/22097603/unit-testing-of-private-functions-with-mocha-and-node-js)

#### Method 1. Use 'rewire' module

```console
npm i --save-dev @types/rewire
```

If you try to import it as `import * as rewire from 'rewire';`, it says
```
This module can only be referenced with ECMAScript imports/exports by turning on the 'esModuleInterop' flag and referencing its default export.ts(2497)
```

We have set esModuleInterop as True. Let's try

```typescript
const rewire = require('rewire');
const ConfigObject = rewire('../../OneExplorer/ConfigObject');
```

#### Method 2. export private functions but distinguish them from public functions by adding underscore _

```typescript


  // Export private functions for unit testing
  static _parseBaseModels = ConfigObj.parseBaseModels;
  static _parseDerivedModels = ConfigObj.parseBaseModels;
```

**Comment by hyunsik-yoon** [comment]
> Even if private methods are covered internally by the module's public api, they are counted as uncovered lines.

/cc @llFreetimell Could you also check this? If this is the case only with typescript (not with other languages, e.g., Java or Python), I guess we'd better discuss this with SE.

**Comment by dayo09** [comment]
FYI,

http://shoulditestprivatemethods.com/ -> NO
https://twitter.com/kentbeck/status/3579860805

People say testing private method is not needed (Kent Beck is a figure in the tdd scene).
They recommend testing only the exposed apis, not wasting time on private functions.

I would be happy if we can skip them. :-D

**Comment by llFreetimell** [comment]
> Even if private methods are covered internally by the module's public api, they are counted as uncovered lines.

When I test without cleaning `/tmp/one-vscode.test` directory, I got above uncovered result.
However, after I cleaned `/tmp/one-vscode.test`, I got following expected result.

File                                         | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s                             
---------------------------------------------|---------|----------|---------|---------|-----------------------------------------------      
ConfigObject.ts          |   76.66 |       50 |      50 |    80.7 | 60-88,174,180,213,221


> They recommend testing only the exposed apis, not wasting time on private functions.

I agree with it but I am not confident to persuade SE team :(

**Comment by dayo09** [comment]
Today I checked the coverage report of https://github.com/Samsung/ONE-vscode/pull/1102 and I could see that private methods are covered.

It seems that when I generated the method, there was an error while running a unittest, because I can see the log as below.

```
Test ConfigObject/2 - Start
Test file is created (/tmp/one-vscode.test/ConfigObject/ConfigObject/2/model.cfg)
Cannot remove the test directory
Test ConfigObject/2 - Done
Error: 1 tests failed.
```

Therefore it seems the line counting is done over private methods, too.
I am closing this issue.
